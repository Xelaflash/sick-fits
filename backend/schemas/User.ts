import { list } from '@keystone-next/keystone/schema';
import { text, password, relationship } from '@keystone-next/fields';

export const User = list({
  // access:
  // ui:
  fields: {
    name: text({ isRequired: true }),
    email: text({ isRequired: true, isUnique: true }),
    password: password(),
    //  TODO: add roles, cart and orders
    cart: relationship({
      ref: 'CartItem.user',
      // Allow to have multiple items in cart
      many: true,
      ui: {
        createView: { fieldMode: 'hidden' },
        itemView: { fieldMode: 'read' },
      },
    }),
    role: relationship({
      ref: 'Role.assignedTo',
      //  TODO: Add access control
    }),
    orders: relationship({ ref: 'Order.user', many: true }),
    products: relationship({ ref: 'Product.user', many: true }),
  },
});
