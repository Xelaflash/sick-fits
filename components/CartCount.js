import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const Dot = styled.div`
  background: var(--red);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  /* below makes all numbers same width */
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.6s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(3.2) rotateY(0.5turn);
  }
  .count-enter-active {
    transform: rotateY(0);
  }
  .count-exit {
    position: absolute;
    top: 0;
    transform: rotateY(0);
  }
  .count-exit-active {
    transform: scale(3.2) rotateY(0.5turn);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 600, exit: 600 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}
