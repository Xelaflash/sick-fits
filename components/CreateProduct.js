import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';
import Form from './styles/Form';

// product mutation is creating a product
const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # which variables are getting passed in along with which type
    # ! in graphQL is to make the var REQUIRED
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      name
      price
      description
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange, clearForm, resetForm } = useForm({
    name: 'prien',
    price: 545,
    description: 'desc test',
  });

  const [createProduct, { loading, error, data }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    // permits to refetch data from backend after product creation
    { variables: inputs, refetchQueries: [{ query: ALL_PRODUCTS_QUERY }] }
  );

  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault();
        //  submit inputs to backend
        const res = await createProduct();
        clearForm();
        // Go to the new product page
        Router.push({
          pathname: `/product/${res.data.createProduct.id}`,
        });
      }}
    >
      <DisplayError error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="image">
          Image
          <input
            required
            type="file"
            id="image"
            name="image"
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">+ Add product</button>
        {/* <button type="button" onClick={clearForm}>
        Clear
        </button>
        <button type="button" onClick={resetForm}>
        Reset
      </button> */}
      </fieldset>
    </Form>
  );
}
