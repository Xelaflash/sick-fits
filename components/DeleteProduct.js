import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

// payload is whats gets returned from the mutation
function update(cache, payload) {
  console.log(payload);
  console.log('running refreshFrontENd function after a delete');
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  //  the delete mutation
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });

  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure you want to DELETE this item?')) {
          // go ahead and actually delete the product
          console.log('deleting');
          deleteProduct().catch((err) => alert(err));
        }
      }}
    >
      {children}
    </button>
  );
}
