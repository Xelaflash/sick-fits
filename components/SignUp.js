import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    createUser(data: { email: $email, name: $name, password: $password }) {
      id
      email
      name
    }
  }
`;

export default function SignUp() {
  const { inputs, handleChange } = useForm({
    email: '',
    name: '',
    password: '',
  });

  const [signup, { data, loading, error }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
  });

  async function handleSubmit(e) {
    e.preventDefault();
    // send email and password to graphQl API
    const res = await signup().catch(console.error);
    console.log(res);
    console.log({ data, loading, error });
  }

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Create an Account</h2>
      <DisplayError error={error} />
      <fieldset
        className="inputs-subgrid"
        disabled={loading}
        aria-busy={loading}
      >
        {data?.createUser && (
          <p>Sign Up with {data.createUser.email} - Please Sign In </p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="x@x.com"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            type="text"
            name="name"
            placeholder="John Doe"
            autoComplete="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            placeholder="Password"
            autoComplete="password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Create Account</button>
      </fieldset>
    </Form>
  );
}

export { SIGNUP_MUTATION };
