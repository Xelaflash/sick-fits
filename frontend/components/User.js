// here we will have a big query to get the current user.
// In order to reuse if needed we will create a custom Hook that returns the current user.

import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';

export const CURRENT_USER_QUERY = gql`
  query {
    # authenticatedItem return a "union" (a type) so need to do the specific syntax in graphql ... on User ==> means give me the datas when authetificatedItem return a User unions.
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            price
            name
            description
            photo {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`;

export function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  return data?.authenticatedItem;
}
