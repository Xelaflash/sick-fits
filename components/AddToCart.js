import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';
import { useCart } from '../lib/cartState';

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export default function AddToCart({ id }) {
  const { openCart } = useCart();
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  return (
    <button
      type="button"
      onClick={() => {
        addToCart();
        setTimeout(() => {
          openCart();
        }, 1700);
      }}
      disabled={loading}
    >
      Add{loading && 'ing'} to Cart ðŸ›’
    </button>
  );
}
