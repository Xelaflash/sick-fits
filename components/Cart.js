import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import { useUser } from './User';
import CartItem from './CartItem';
import formatMoney from '../lib/formatMoney';
import calcTotalCartPrice from '../lib/calcTotalCartPrice';
import { useCart } from '../lib/cartState';
import CloseButton from './styles/CloseButton';
import { Checkout } from './Checkout';

export default function Cart() {
  const currentUser = useUser();
  const { cartOpen, closeCart } = useCart();
  // console.log(cartOpen);
  if (!currentUser) return null;
  // console.log(currentUser);
  return (
    <CartStyles open={cartOpen}>
      <CloseButton type="button" onClick={closeCart} id="close-cart">
        &times;
      </CloseButton>
      <header>
        <Supreme>{currentUser.name}</Supreme>
      </header>
      <ul>
        {currentUser.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalCartPrice(currentUser.cart))}</p>
        <Checkout />
      </footer>
    </CartStyles>
  );
}
