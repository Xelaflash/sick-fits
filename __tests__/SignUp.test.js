import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const user = fakeUser();

const mocks = [
  // mutation mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: user.name,
        email: user.email,
        password: 'pwd',
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'idofuser',
          name: user.name,
          email: user.email,
        },
      },
    },
  },
  // current user mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: user } },
  },
];

describe('<SignUp />', () => {
  it('1. renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('2. Call mutation properly ', async () => {
    const { debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );
    // type into input fields
    await userEvent.type(screen.getByPlaceholderText('John Doe'), user.name);
    await userEvent.type(screen.getByPlaceholderText('x@x.com'), user.email);
    await userEvent.type(screen.getByPlaceholderText('Password'), 'pwd');
    // submit event
    await userEvent.click(screen.getByText('Create Account'));
    await screen.findByText(`Sign Up with ${user.email} - Please Sign In`);
    // debug();
  });
});
