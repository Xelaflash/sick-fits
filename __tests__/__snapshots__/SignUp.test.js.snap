// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> 1. renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 iaUpjc"
    method="POST"
  >
    <h2>
      Create an Account
    </h2>
    <fieldset
      aria-busy="false"
      class="inputs-subgrid"
    >
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="x@x.com"
          type="email"
          value=""
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          name="name"
          placeholder="John Doe"
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Create Account
      </button>
    </fieldset>
  </form>
</div>
`;
